src/codegen/codegen.cc

```diff
@@ -58,20 +60,110 @@ runtime::Module Build(const Array<LoweredFunc>& funcs,
   return m;
 }
 
+/*! \brief Helper class to serialize module */
+class ModuleSerializer {
+ public:
+  explicit ModuleSerializer(runtime::Module mod) : mod_(mod) {
+    Init();
+  }
+
+  void SerializeModule(dmlc::Stream* stream) {
+    // Only have one DSO module and it is in the root, then
+    // we will not produce import_tree_.
+    bool has_import_tree = true;
+    if (IsDSOModule(mod_->type_key()) && mod_->imports().empty()) {
+      has_import_tree = false;
+    }
+    uint64_t sz = 0;
+    if (has_import_tree) {
+      // we will append one key for _import_tree
+      // The layout is the same as before: binary_size, key, logic, key, logic...
+      sz = mod_vec_.size() + 1;
+    } else {
+      // Keep the old behaviour
+      sz = mod_->imports().size();
+    }
+    stream->Write(sz);
+
+    for (auto m : mod_vec_) {
+      std::string mod_type_key = m->type_key();
+      if (!IsDSOModule(mod_type_key)) {
```

Comment:
I think it is cleaner to let IsDSOModule take module as the name also suggests it

```c++
bool IsDSOModule(const Module& mod) const {
    return mod->type_key() == "llvm" || mod->type_key() == "c";
}
```

and then we don't really need `mod_type_key` here

Reply:
The function name could be changed. But the mod_type_key is needed. Because when we deserialize, we assume the type is std::string, but mod->type_key() returns const char*